### PR TITLE
Fix incorrect data interval alignment due to assumption on input time alignment

### LINKED ISSUE #23689: Data Interval wrong when manually triggering with a specific logical date
### Apache Airflow version

2.2.5

### What happened

When I use the date picker in the “Trigger DAG w/ config” page to choose a specific logical date for some reason on a scheduled daily DAG the Data Interval Start (circled in red) is 2 days before the logical date (circled in blue), instead of the same as the logical date. And the Data Interval End is one day before the logical date. So the interval is the correct length, but on wrong days.

Screen Shot 2022-05-11 at 5 14 10 PM

I encountered this with a DAG with a daily schedule which typically runs at 09:30 UTC. I am testing this in a dev environment (with catchup off) and trying to trigger a run for 2022-05-09 09:30:00. I would expect the data interval to start at that same time and the data interval end to be 1 day after.

It has nothing to do with the previous run since that was way back on 2022-04-26

### What you think should happen instead

The data interval start date should be the same as the logical date (if it is a custom logical date)

### How to reproduce

I made a sample DAG as shown below:
```python
import pendulum
from airflow.models import DAG
from airflow.operators.python import PythonOperator


def sample(data_interval_start, data_interval_end):
    return "data_interval_start: {}, data_interval_end: {}".format(str(data_interval_start), str(data_interval_end))


args = {
    'start_date': pendulum.datetime(2022, 3, 10, 9, 30)
}
with DAG(
    dag_id='sample_data_interval_issue',
    default_args=args,
    schedule_interval='30 9 * * *'  # 09:30 UTC
) as sample_data_interval_issue:
    task = PythonOperator(
        task_id='sample',
        python_callable=sample
    )

```

I then start it to start a scheduled DAG run (`2022-05-11, 09:30:00 UTC`), and the `data_interval_start` is the same as I expect, `2022-05-11T09:30:00+00:00`.

However, when I went to "Trigger DAG w/ config" page and in the date chooser choose `2022-05-09 09:30:00+00:00`, and then triggered that. It shows the run datetime is `2022-05-09, 09:30:00 UTC`, but the `data_interval_start` is incorrectly set to `2022-05-08T09:30:00+00:00`, 2 days before the date I choose.

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

N/A

### Deployment

Astronomer

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by wanlce**
This introduces a new `_align_to_prev` util function for cron-based interval alignment. Previously we assumed `_get_prev(_align_to_next(t))` is enough for this purpose, but this proved incorrect (see discussions in #21715), and an additional alignment logic is needed to avoid the inferred interval being off by one period when the input value has unexpected alignment (unexpectedly aligning for a manual run, and unexpected _not aligning_ for an auto one).

See also #21715 
Fix #23689

### CODE REVIEW

**Comment by wanlce** [comment]
I am very sorry, because my wrong rebase caused the old request to be closed. :sob: 
I wasn't sure where these changes needed to be documented, thank you so much for your help!

File: airflow/timetables/interval.py [hunk 1]
```diff
@@ -58,6 +58,10 @@ def _align(self, current: DateTime) -> DateTime:
         """
         raise NotImplementedError()
 
+    def _align_to_prev(self, current: DateTime) -> DateTime:
+        """Align given time to the scheduled."""
```
Comment (uranusjr) [inline airflow/timetables/interval.py]: Please

1. Add more description on what this function does (similar to the docstring for `_align_to_next`)
2. Describe _why_ this is needed and how it’s different from `self._prev(self._align_to_next(t))` (when a DAG’s schedule changes, as you mentioned in the original PR).

File: airflow/timetables/interval.py [hunk 2]
```diff
@@ -76,21 +80,23 @@ def next_dagrun_info(
...
-            earliest = self._align(earliest)
+            earliest = self._align_to_next(earliest)
         if last_automated_data_interval is None:
             # First run; schedule the run at the first available time matching
             # the schedule, and retrospectively create a data interval for it.
             if earliest is None:
                 return None
             start = earliest
-        else:  # There's a previous run.
+        else:  # There's a previous run
```
Comment (uranusjr) [inline airflow/timetables/interval.py]: Accidental?

Comment (wanlce) [reply airflow/timetables/interval.py]: sorry !!!   Thank you for your review, I made changes. Looking forward to your new comment.

Comment (wanlce) [reply airflow/timetables/interval.py]: Hello

File: airflow/timetables/interval.py [hunk 3]
```diff
@@ -58,6 +58,21 @@ def _align(self, current: DateTime) -> DateTime:
...
+
+        For fixed schedules (e.g. every midnight); this finds the prev time that
+        aligns to the declared time, if the given time does not align. If the
+        schedule is not fixed (e.g. every hour), the given time is returned.
+
+        It is not enough to just have _align_to_next. When a DAG's schedule
+        changed, _algin_to_next does not correct the scheduling time in time,
+        resulting in the first scheduling after the scheduling time has been
+        changed remaining the same. So we would need to align forward to ensure
+        that it works correctly after a scheduling time change.
```
Comment (uranusjr) [inline airflow/timetables/interval.py]:
```suggestion
        It is not enough to use ``_get_prev(_align_to_next())`` instead. When a
        DAG's schedule changes, this alternative would make the first scheduling
        after the schedule change remain the same. We need to align forward to
        ensure it works correctly in this situation.
```

**Review by uranusjr — CHANGES_REQUESTED** [review]

**Review by uranusjr — APPROVED** [review]

**Comment by uranusjr** [comment]
Anyone else wants to take a look at this one? It’s a relatively straightforward change, since we established that this additional method is needed.

**Comment by potiuk** [comment]
I think  what woudl be great here is to add a few more tests showing the different cases  and maybe explaining the context in the test a bit better?

The case here is either not tested or it's not clear that it is tested I think: 

```
    def _align_to_prev(self, current: DateTime) -> DateTime:
        """Get the prev scheduled time.
        This is ``current - interval``, unless ``current`` falls right on the
        interval boundary, when ``current`` is returned.
        """
        prev_time = self._get_prev(current)
        if self._get_next(prev_time) != current:
            return prev_time
        return current
```

**Comment by uranusjr** [comment]
Actually this is also the cause of #23689. I’ll add a few test cases here to illustrate those isues.

**Review by ashb — APPROVED** [review]

**Pull request merged**